src/lib.rs

```diff
@@ -33,6 +33,8 @@ pub(crate) type Result<T> = result::Result<T, CollectionError>;
 pub struct CrateInfo<'a> {
     pkg_name: &'a str,
     pkg_version: &'a str,
+    #[cfg(feature = "git_hash")]
+    git_hash: &'a str,
```

Comment:
What do you think of the following idea:

Instead of adding feature flags all over the place, we:
- Always add a `git_hash` field, but make it a `Option<&'a str>`.
- Leave the `from_name_and_version` constructor unchanged, but add a `set_git_hash` method on `BugReport`
- Have just a single version of the `bugreport!` macro, and add something like
  ```rust
  #cfg(feature = "git_hash")]
  {
    report.set_git_hash(…);
  }
  ```
  ```

Reply:
In the case of `Option<&'a str>` I had a that first too, but thought going full `cfg` ended up simpler. But I’ll change back to `Option<&'a str>` so you can decide for yourself which you prefer, no problem at all. I agree there is a neatness to not have `cfg` all over the place.

At first I had `cfg` inside of the `bugreport!()` macro. But when testing with bat, realized that it will not work, since the `cfg` will be evaluated in the context of the dependent project (e.g. bat), for which the feature is not enabled or even defined. I wrote a note about it in the commit (and PR) message.

Admittedly, I did not explore in depth ways to get around that, so I’ll see if I can figure out a way.

Thanks for the review!

Reply:
> In the case of `Option<&'a str>` I had a that first too, but thought going full `cfg` ended up simpler. But I’ll change back to `Option<&'a str>` so you can decide for yourself which you prefer, no problem at all. I agree there is a neatness to not have `cfg` all over the place.

:+1: 



> At first I had `cfg` inside of the `bugreport!()` macro. But when testing with bat, realized that it will not work, since the `cfg` will be evaluated in the context of the dependent project (e.g. bat), for which the feature is not enabled or even defined. I wrote a note about it in the commit (and PR) message.

Sorry, I completely forgot to read the PR description :smile:. That makes total sense. Thanks for documenting it properly!

> Admittedly, I did not explore in depth ways to get around that, so I’ll see if I can figure out a way.

Not necessary.

Reply:
I changed to use `Option<&'a str>` in dec5a3e  that I just pushed to this MR, and now when I see it complete in all its glory I prefer it over using `cfg` :)

In ae76ca1 I tried a new approach to using different `bugreport!()` definitions. I now simply use a (hidden) helper macro. I prefer this over the old approach, since it scales better. Imagine if we wanted to introduce a 2 new configurable features with similar implementation constraints. Then we would need 8 different definitions of the `bugreport!()` macro :fearful:  So I think the new approach is more future proof.

Looking forward to your comments on the new code! 

(Btw, I advice to check the full diff when reviewing, rather than diff of individual commits, since the back-and-forth is a bit noisy. The full diff is much more simple and clear I think.)

Reply:
> In [ae76ca1](https://github.com/sharkdp/bugreport/commit/ae76ca1836eb558a5b3b467c5a0648904b052cf1) I tried a new approach to using different `bugreport!()` definitions. I now simply use a (hidden) helper macro. I prefer this over the old approach, since it scales better. Imagine if we wanted to introduce a 2 new configurable features with similar implementation constraints. Then we would need 8 different definitions of the `bugreport!()` macro fearful So I think the new approach is more future proof.

Nice :+1: